### PR TITLE
Add toggle suggestion details for link autocomplete

### PR DESCRIPTION
**Pull request by ahazxm**
#38 
The key idea to add suggestion details is to set `CompletionItem.documentation` to the content of the corresponding file, for doing that I have to `openTextDocument` for each file.

The current implementation would make auto-completion not so fast, so I also add a configuration (`provideSuggestionDetails`) to turn off suggestion details (the default value is **false**)

would be happy to discuss more efficient methods with you.

Thank you for this incredible extension!

### CODE REVIEW

**Comment by kortina** [comment]
OK, this will be very sweet and I like the sway you added a configuration option for it.

AND, I think it won't be that hard to make it more performant.

We already have support for caching the results of expensive parsing operations for notes, and I think it won't be hard to integrate our code to that. The relevant section is the `Note` class in `NoteParser`:

https://github.com/kortina/vscode-markdown-notes/blob/2cb1262cca8fa218fe6d38256a64fc3146983722/src/NoteParser.ts#L55

I *think* here where we are constructing the label:

https://github.com/kortina/vscode-markdown-notes/pull/65/files#diff-c83483a19101a6a90334944e70c93ad5L41

We *should* be able to get the cached note from `NoteParser._notes` -- currently private dict indexed by the `fsPath` of the note, which you will have as `f.fsPath`

I might suggest adding a methods:

- `public NoteParser.noteFromFsPath` - return the `Note` object from the cache
- `public Note.completionDocumentation` - move your code `item.documentation = new vscode.MarkdownString(doc.getText());` / `item.documentation = doc.getText();` in here and use the `data` cache of the doc text instead
- `public Note.completionItemDetails` - return the `{kind, label, range, documentation}` you want
- `public Note.title` - this is a bonus and probably out of scope but if you are feeling ambitious it would help out with some other requests -- cache the first `# h1` text found in the note so we have an in-memory store of the title of the note. This would be useful for allowing completions in "Title Case" instead of "kebab-case"

Does that all make sense? 

Lmk if you have any questions.

Excited about this improvement. Tx for contributing.

**Comment by ahazxm** [comment]
A better understanding of the code! Thank you!

I'll give it a try.

**Comment by ahazxm** [comment]
I remove the configuration `provideSuggestionDetails` since `.documentation` doesn't affect the efficiency of auto-completion anymore.

**Comment by ahazxm** [comment]
cache the first `# h1` here:

https://github.com/ahazxm/vscode-markdown-notes/blob/f9a3b0f75ef5ff7a250a16bcb5cc5b322318587d/src/NoteParser.ts#L131-L139

Thx for your useful suggestions!

**Comment by lukesmurray** [comment]
this is a great idea @ahazxm. One thing you may be able to do to improve performance is implement [resolveCompletionItem](https://code.visualstudio.com/api/references/vscode-api#CompletionItemProvider.resolveCompletionItem) in the completion provider. That way you don't have to fill in details or documentation until it is needed in the UI.

The one caveat is you need some way to resolve a completion item to the original Note. I'm not sure if there is an easy way to do that here. In my code I can convert from a label to a note so its not too hard. You could potentially pass the `fspath` in the completion item somehow and then this would be trivial with the `noteFromFsPath` @kortina described.

Here's an example from my code which fills in documentation based on the label.

```ts
  public async resolveCompletionItem(
    item: vscode.CompletionItem,
    token: vscode.CancellationToken
  ): Promise<vscode.CompletionItem> {
    // get the fsPath from the label
    const fsPath = getCurrentFSPathForTitle(this.store.getState())(item.label);
    // convert the fsPath to a markdown string
    item.documentation = await fsPathMarkdownString(fsPath, this.store);
    return item;
  }
```

**Comment by kortina** [comment]
Looking good, @ahazxm. I will try to review the code more thoroughly in the next few days. Excited for this!

**Comment by ahazxm** [comment]
@lukesmurray  great suggestion! But I found now it's hard to directly get the **correct** `fsPath` of the selected completion.

The key issue is when setting `useUniqueFilenames`, a note `one.md` in dir `subfolder` can link as `[[one]]` without the dir `subfolder`, here is an example:
```markdown
This is a note in the root directory.
[[one]] is a note in directory *subfolder*
``` 
So it's hard to know the folder it belongs to and the `fsPath` directly from the `label` (or the formatted filename) like
```typescript
const fsPath = getCurrentFSPathForTitle(this.store.getState())(item.label);
```
The key is `this.store`?

Now to get the `fsPath` we may need to do `fuzzyMatch` like the method in `MarkdownDefinitionProvider`
https://github.com/kortina/vscode-markdown-notes/blob/2cb1262cca8fa218fe6d38256a64fc3146983722/src/MarkdownDefinitionProvider.ts#L30-L43

or a new dict with `key: label` and `value: fsPath`?

would like to have a more in-depth discussion with you.

**Comment by ahazxm** [comment]
@kortina I'm trying to implement [resolveCompletionItem](https://code.visualstudio.com/api/references/vscode-api#CompletionItemProvider.resolveCompletionItem) @lukesmurray mentioned. 

When there are a large number of notes and `completionItems`, it is really essential to set up `documentation` only when the item is selected.

**Comment by kortina** [comment]
@ahazxm I'm sorry, I completely lost track of this conversation and your question.

Let me try to restate to make sure I am understanding correctly.

You would like to implement [resolveCompletionItem](https://code.visualstudio.com/api/references/vscode-api#CompletionItemProvider.resolveCompletionItem) in `MarkdownFileCompletionItemProvider` https://github.com/ahazxm/vscode-markdown-notes/blob/f9a3b0f75ef5ff7a250a16bcb5cc5b322318587d/src/MarkdownFileCompletionItemProvider.ts#L12

Inside this method, when `useUniqueFilenames=true` (the default setting), you need to find the correct `fsPath` for the completion item in question?

I think there are 2 options.

First, I see that we do have the correct `fsPath` in `provideCompletionItems`: https://github.com/ahazxm/vscode-markdown-notes/blob/f9a3b0f75ef5ff7a250a16bcb5cc5b322318587d/src/MarkdownFileCompletionItemProvider.ts#L43

Could you subclass `CompletionItem`, eg, `CompletionItemWithFsPath`, and just add an attribute `fsPath` to the item here, so then you'll have it downstream when `resolveCompletionItem` gets called? Doing something like this seems ideal.

If that does not work, then inside `resolveCompletionItem` you could, as you suggested, basically call the same code you reference from MarkdownDefinitionProvider`.

If you go this latter route, what I might suggest is to factor out everything up until the create-if-missing-section:
https://github.com/ahazxm/vscode-markdown-notes/blob/f9a3b0f75ef5ff7a250a16bcb5cc5b322318587d/src/MarkdownDefinitionProvider.ts#L56

into a separate, `provideDefinitionButDontCreateIfMissing` or something like that, call that inside `provideDefinition`, and then you should be able to use `provideDefinitionButDontCreateIfMissing` directly from your new `resolveCompletionItem` function.

Does that all make sense? If not, happy to discuss more / clarify.

Sorry for the slow reply on this! I'll turn around answers much faster if you have more questions.

**Comment by ahazxm** [comment]
Hi, @kortina, a subclass of `CompletionItem` is a great idea, I extend the `CompletionItem` here:
https://github.com/ahazxm/vscode-markdown-notes/blob/41d5f0b9d9cefc3295a71de8b6a75b15d9be0abe/src/MarkdownFileCompletionItemProvider.ts#L6

now inside `resolveCompletionItem` we can easily obtain the `fsPath` of each item (https://github.com/ahazxm/vscode-markdown-notes/blob/41d5f0b9d9cefc3295a71de8b6a75b15d9be0abe/src/MarkdownFileCompletionItemProvider.ts#L56)

thx for your useful suggestion :) 👍

File: src/MarkdownFileCompletionItemProvider.ts [hunk 1]
```diff
@@ -44,4 +52,17 @@ export class MarkdownFileCompletionItemProvider implements vscode.CompletionItem
...
     }
   }
+
+  public async resolveCompletionItem(
+    item: MarkdownFileCompletionItem,
+    token: vscode.CancellationToken
+  ): Promise<MarkdownFileCompletionItem> {
+    const fsPath = item.fsPath;
+    if (fsPath) {
+      let note = NoteParser.noteFromFsPath(fsPath);
```
Comment (kortina) [inline src/MarkdownFileCompletionItemProvider.ts:62]: I see below the return type of this function is `Note`, but I wonder if it should be `Note | undefined`.

Generally `NoteWorkspace._notes[fsPath]` *should* exist, but I worry there may be race conditions where if the timing of the call to this just happens to be a moment when this value has been cleared / before it has been set, we might return `undefined`

So I think we might want to modify the return type of the function and add handling to the callers (like this line) to handle the case where `note = undefined`

File: src/NoteParser.ts [hunk 2]
```diff
@@ -177,6 +186,17 @@ export class Note {
       });
     return _tagSet;
   }
+
+  // completionItem.documentation ()
+  documentation(): string | vscode.MarkdownString | undefined {
+    if (this.data === undefined) {
+      return "";
+    } else if (NoteWorkspace.compileSuggestionDetails()) {
+      return new vscode.MarkdownString(this.data);
```
Comment (kortina) [inline src/NoteParser.ts:195]: Do we need to wrap the call to `MarkdownString` in a try catch? Could it raise an exception for any weird string value of `this.data` ?

File: src/NoteWorkspace.ts [hunk 3]
```diff
@@ -51,6 +51,7 @@ export class NoteWorkspace {
   static _rxTagNoAnchors = '\\#[\\w\\-\\_]+'; // used to match tags that appear within lines
   static _rxTagWithAnchors = '^\\#[\\w\\-\\_]+$'; // used to match entire words
   static _rxWikiLink = '\\[\\[[^sep\\]]+(sep[^sep\\]]+)?\\]\\]'; // [[wiki-link-regex(|with potential pipe)?]] Note: "sep" will be replaced with pipedWikiLinksSeparator on compile
+  static _rxTitle = '(?<=^( {0,3}#[^\\S\\r\\n]+))([^\\r\\n]+)';
```
Comment (kortina) [inline src/NoteWorkspace.ts:54]: Can you add a comment that explains a bit about what this regex matches? 

I am pretty familiar with regex, but tbh, I don't quite get what all the characters before ` {0,3}#` are doing. Is this 0 to 3 space characters before a `#` ? What is the `(?<=^(` ?

And so the `#` is directly succeed by 1 or more characters that not not-space, return, or newline? I don't get why the `\\S` is in there, actually?

File: package.json [hunk 4]
```diff
@@ -95,6 +95,11 @@
           "default": "# ${noteName}\n\n",
           "description": "Template for auto-created note files. Available tokens: ${noteName}, ${timestamp}. Timestamp is inserted in ISO format, i.e. 2020-07-09T05:29:00.541Z."
         },
+        "vscodeMarkdownNotes.compileSuggestionDetails": {
```
Comment (kortina) [inline package.json:98]: What are the cases where someone would set this to false? If they just use `plaintext` notes?

Is there a perf implication?

Also, can you explain a little of your thinking around the `false` default? Generally that is my bias, but curious to hear both sides of this and could imagine this being true if there is not a perf reason and you think most people would want this on

File: src/MarkdownFileCompletionItemProvider.ts (hunk 1 above)
Comment (ahazxm) [reply src/MarkdownFileCompletionItemProvider.ts:62]: It should be `Note | undefined`, I will fix it.

File: src/NoteParser.ts (hunk 2 above)
Comment (ahazxm) [reply src/NoteParser.ts:195]: Actually I'm not sure if an exception will be raised, but adding a `try catch` shouldn't be a problem and I'll fix it.

File: src/NoteWorkspace.ts (hunk 3 above)
Comment (ahazxm) [reply src/NoteWorkspace.ts:54]: I make an explanation in [here](https://regex101.com/r/g89Y3U/2) :)

File: package.json (hunk 4 above)
Comment (ahazxm) [reply package.json:98]: I forgot to change this default.

the fact is that I don't really like the display of markdownString, the size of the title is really unattractive, so I added this configuration, and if anyone who doesn't like this,  can also choose to set this option to false, but the default value should really be `true`!

Screen Shot 2020-08-21 at 4 01 55 PM
Screen Shot 2020-08-21 at 4 02 54 PM

File: src/NoteWorkspace.ts (hunk 3 above)
Comment (ahazxm) [reply src/NoteWorkspace.ts:54]: I'll submit changes after we finish our discussion and work out all the issues

I am looking forward to your reply :)

**Pull request merged**